File: ios/ConstructionCalc/Model/Preferences.swift
import Foundation
import CalcEngine

/// Which length format the calculator opens in. Raw values match the
/// `DefaultLengthFormat` string union in `frontend/src/lib/preferences.ts`
/// so a tape / preference exported from the web reads identically here.
enum LengthFormatPref: String, CaseIterable, Identifiable {
    case feetInchFraction = "feet_inch_fraction"
    case decimalFeet = "decimal_feet"
    case decimalInches = "decimal_inches"
    case meters
    case yards

    var id: String { rawValue }

    var label: String {
        switch self {
        case .feetInchFraction: return "Feet + inch + fraction"
        case .decimalFeet: return "Decimal feet"
        case .decimalInches: return "Decimal inches"
        case .meters: return "Meters"
        case .yards: return "Yards"
        }
    }
}

/// Allowed rounding resolutions for the feet-inch-fraction display. Matches
/// `FractionDenom = 4 | 8 | 16` on the web and the engine's
/// `denom ∈ {2,4,8,16}` validation.
enum FractionDenomPref: Int, CaseIterable, Identifiable {
    case quarter = 4
    case eighth = 8
    case sixteenth = 16

    var id: Int { rawValue }
    var label: String { "1/\(rawValue)\"" }
}

/// User preferences, persisted to `UserDefaults`. The web equivalent is
/// `localStorage` (`preferences.ts`); the shape is intentionally identical
/// so the two stay in lockstep.
struct Preferences {
    var fractionDenom: FractionDenomPref
    var lengthFormat: LengthFormatPref
    var angleInDegrees: Bool

    static let defaults = Preferences(
        fractionDenom: .sixteenth,
        lengthFormat: .feetInchFraction,
        angleInDegrees: true
    )

    /// UserDefaults keys. Kept here so the view model and the preferences
    /// screen read/write the same storage.
    enum Key {
        static let denom = "cc.pref.fractionDenom"
        static let format = "cc.pref.lengthFormat"
        static let degrees = "cc.pref.angleInDegrees"
    }

    static func load(from defaults: UserDefaults = .standard) -> Preferences {
        var prefs = Preferences.defaults

        if let raw = defaults.object(forKey: Key.denom) as? Int,
           let denom = FractionDenomPref(rawValue: raw) {
            prefs.fractionDenom = denom
        }
        if let raw = defaults.string(forKey: Key.format),
           let format = LengthFormatPref(rawValue: raw) {
            prefs.lengthFormat = format
        }
        if defaults.object(forKey: Key.degrees) != nil {
            prefs.angleInDegrees = defaults.bool(forKey: Key.degrees)
        }
        return prefs
    }

    func save(to defaults: UserDefaults = .standard) {
        defaults.set(fractionDenom.rawValue, forKey: Key.denom)
        defaults.set(lengthFormat.rawValue, forKey: Key.format)
        defaults.set(angleInDegrees, forKey: Key.degrees)
    }

    /// Translate the chosen format into the engine `Convert` event. Mirrors
    /// `preferencesToConvertKey()` in `preferences.ts`.
    func toConvertKey() -> KeyEvent {
        switch lengthFormat {
        case .feetInchFraction:
            return .convert(format: .feetInchFraction(denom: UInt32(fractionDenom.rawValue)))
        case .decimalFeet:
            return .convert(format: .decimalFeet(precision: 4))
        case .decimalInches:
            return .convert(format: .decimalInches(precision: 4))
        case .meters:
            return .convert(format: .meters(precision: 4))
        case .yards:
            return .convert(format: .yards(precision: 4))
        }
    }

    /// Translate the angle preference into the engine `SetAngleMode` event.
    /// Mirrors `preferencesToAngleModeKey()` in `preferences.ts`.
    func toAngleModeKey() -> KeyEvent {
        .setAngleMode(degrees: angleInDegrees)
    }
}

File: frontend/src/lib/preferences.ts
/**
 * User preferences, persisted to localStorage.
 *
 * The schema is versioned via the storage key so a future breaking
 * change (`v1` → `v2`) just falls back to defaults instead of crashing
 * on an old object shape. iOS will mirror this on UserDefaults using
 * the same Preferences shape.
 */

import type { Key } from './calc';

const STORAGE_KEY = 'cc.preferences.v1';

export type FractionDenom = 4 | 8 | 16;

export type DefaultLengthFormat =
  | 'feet_inch_fraction'
  | 'decimal_feet'
  | 'decimal_inches'
  | 'meters'
  | 'yards';

export interface Preferences {
  /** Rounding resolution for the feet-inch-fraction display. */
  defaultFractionDenom: FractionDenom;
  /** Which length format the calculator opens in. */
  defaultLengthFormat: DefaultLengthFormat;
  /** Trig keys interpret/return angles in degrees when true, radians when false. */
  angleInDegrees: boolean;
}

export const DEFAULTS: Preferences = {
  defaultFractionDenom: 16,
  defaultLengthFormat: 'feet_inch_fraction',
  angleInDegrees: true
};

export function loadPreferences(): Preferences {
  if (typeof localStorage === 'undefined') return { ...DEFAULTS };
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return { ...DEFAULTS };
    const parsed = JSON.parse(raw);
    return {
      defaultFractionDenom: coerceDenom(parsed.defaultFractionDenom),
      defaultLengthFormat: coerceFormat(parsed.defaultLengthFormat),
      angleInDegrees: typeof parsed.angleInDegrees === 'boolean' ? parsed.angleInDegrees : true
    };
  } catch {
    // Corrupt JSON — fall back rather than wedge the app forever.
    return { ...DEFAULTS };
  }
}

export function savePreferences(p: Preferences): void {
  if (typeof localStorage === 'undefined') return;
  localStorage.setItem(STORAGE_KEY, JSON.stringify(p));
}

function coerceDenom(v: unknown): FractionDenom {
  return v === 4 || v === 8 || v === 16 ? v : DEFAULTS.defaultFractionDenom;
}

function coerceFormat(v: unknown): DefaultLengthFormat {
  const allowed: DefaultLengthFormat[] = [
    'feet_inch_fraction',
    'decimal_feet',
    'decimal_inches',
    'meters',
    'yards'
  ];
  return allowed.includes(v as DefaultLengthFormat)
    ? (v as DefaultLengthFormat)
    : DEFAULTS.defaultLengthFormat;
}

/**
 * Translate the chosen format into the `Convert` KeyEvent the engine
 * accepts. Pulled out of `+page.svelte` so the iOS app can reuse it
 * when it ships.
 */
export function preferencesToConvertKey(p: Preferences): Key {
  switch (p.defaultLengthFormat) {
    case 'feet_inch_fraction':
      return {
        type: 'convert',
        format: 'feet_inch_fraction',
        denom: p.defaultFractionDenom
      };
    case 'decimal_feet':
      return { type: 'convert', format: 'decimal_feet', precision: 4 };
    case 'decimal_inches':
      return { type: 'convert', format: 'decimal_inches', precision: 4 };
    case 'meters':
      return { type: 'convert', format: 'm', precision: 4 };
    case 'yards':
      return { type: 'convert', format: 'yards', precision: 4 };
  }
}

/**
 * Translate the angle preference into the engine `SetAngleMode` event.
 * Sent alongside `preferencesToConvertKey` at launch so trig keys honour
 * the stored degrees/radians choice.
 */
export function preferencesToAngleModeKey(p: Preferences): Key {
  return { type: 'setAngleMode', degrees: p.angleInDegrees };
}
